milatools/utils/compute_node.py

```diff
@@ -0,0 +1,512 @@
+
+    This handles the case where an interrupt is raised while running a command over SSH
+    that creates a new job allocation (either salloc or sbatch) before we are able to
+    parse the job id. (In the case where we have the job ID, we simply use
+    `scancel {job_id}`).
+
+    In this case, we try to cancel the new job(s) that have appeared since entering the
+    `async with` block that have the name `job_name`. Emits a warning in the (unlikely)
+    case where such jobs are not found, as it means that there could be a "zombie" job
+    allocation on the cluster.
+    """
+    jobs_before = await get_queued_milatools_job_ids(login_node, job_name=job_name)
+    if jobs_before:
+        logger.info(
+            f"Existing jobs on {login_node.hostname} with name {job_name}: {jobs_before}"
+        )
+    else:
+        logger.debug(
+            f"There are currently no jobs with name {job_name} on {login_node.hostname}."
+        )
+    try:
+        yield
+    except (KeyboardInterrupt, asyncio.CancelledError):
+        jobs_after = await get_queued_milatools_job_ids(login_node, job_name=job_name)
+        logger.warning("Interrupted before we were able to parse a job id!")
+        # We were unable to get the job id, so we'll try to cancel only the newly
+        # spawned jobs from this user that match the set name.
+        new_jobs = list(set(jobs_after) - set(jobs_before))
+        if len(new_jobs) == 1:
+            job_id = new_jobs[0]
+            console.log(
+                f"Cancelling job {job_id} since it is the only new job of this "
+                f"user with name {job_name!r} since the last call to salloc or sbatch.",
+                style="yellow",
+            )
+            login_node.run(f"scancel {new_jobs[0]}", display=True, hide=False)
+        elif len(new_jobs) > 1:
+            console.log(
+                f"There appears to be more than one new jobs from this user with "
+                f"name {job_name!r} since the initial call to salloc or sbatch: "
+                f"{new_jobs}\n"
+                "Cancelling all of them to be safe...",
+                style="yellow",
+            )
+            await asyncio.shield(
+                login_node.run_async(
+                    "scancel " + " ".join(str(job_id) for job_id in new_jobs),
+                    display=True,
+                    hide=False,
+                )
+            )
+        else:
+            warnings.warn(
+                RuntimeWarning(
+                    f"Unable to find any new job IDs with name {job_name!r} since the last "
+                    f"job allocation. This means that if job allocations were created, "
+                    "they might not have been properly cancelled. Please check that there "
+                    f"are no leftover jobs with {job_name!r} on the cluster!"
+                )
+            )
+        raise
+
+
+async def salloc(
```

Comment:
If the user loses their internet connection or turns off their laptop, the `salloc` job will end and prevent wasting ressources.

Reply:
> I'm not entirely sure that stopping the job on connection lost is a feature ;) 

Oh but it really is - in my opinion - since we can always have almost instantaneous access to a GPU with the unkillable partition, the downside of having to reconnect is definitely worth the upside of not potentially wasting a GPU for a week if someone disconnects weirdly from a batch job.

Reply:
Definitely keeping both sbatch and salloc for now, at the very least in order to retain backward compatibility with current commands.